content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt

```diff
@@ -0,0 +1,301 @@
+
+While the code we need to generate the ToCs from the headlines in a textarea field is the same, we can use different approaches to add them to the page.
+
+- Use a placeholder that users can manually put into a textarea field where they want the ToC to appear
+- Add a ToC in our template code using a field method
+
+We will look into both options in this recipe.
+
+<info>
+To keep it simple, we will include only one headline level in our ToC. We will cover multiple levels in a future recipe.
+</info>
+
+## Using a `toc` placeholder
+
+For the `toc` placeholder, we use another `kirbyText:after` hook, and a snippet that will contain the HTML for the ToC. We will reuse this snippet for the (link: #alternative-a-field-method-to-generate-the-toc text: alternative field method approach) below.
+
+```php "/site/plugins/toc/index.php"
+<?php
+
+use Kirby\Toolkit\Collection;
+use Kirby\Toolkit\Str;
+use Kirby\Toolkit\Obj;
+
+Kirby::plugin('k-cookbook/toc', [
+  'hooks' => [
+  
+    'kirbytext:after' => [
+
+      // (…) add the callback function from above here (without the wrapper)
+        
+      function($text) {
+
+        // the pattern allows passing an optional headline level `(toc: h3)`
+        $pattern  = '!\(toc:?\s?(h[1-6])?\)!';
+        
+        $text = preg_replace_callback($pattern, function($match) use($text) {
+          
+          // get the headline level from the match
+          $headline = $match[1] ?? 'h2';
+
+          // find all headline matches
+          preg_match_all('!<' . $headline . '.*?>(.*?)</' . $headline . '>!s', $text, $matches);
+
+          // create a new collection for the headlines…
+          $headlines = new Collection;
+    
+          // …and add all matches
+          foreach ($matches[1] as $text) {
+              $headline = new Obj([
+                  'id'   => $id = '#' . Str::slug(Str::unhtml($text)),
+                  'url'  => $id,
+                  'text' => trim(strip_tags($text)),
+              ]);
+
+              $headlines->append($headline->url(), $headline);
+          }
+
+          // return the html for the ToC
+          return snippet('toc', ['headlines' => $headlines], false);
+
+        }, $text);
+    
+        return $text;
+
+      },
+    ],
+  ],
+  'snippets' => [
+    'toc' => __DIR__ . '/snippets/toc.php'
+  ],
+]);
+```
+
+The code for the snippet goes into `/site/plugins/toc/snippets/toc.php`:
+
+```php "/site/plugins/toc/snippets/toc.php"
+<?php if ($headlines->isNotEmpty()) : ?>
+  <nav class="toc">
+
+    <h2>Table of Contents</h2>
+
+    <ol>
+      <?php foreach($headlines as $headline): ?>
+        <li><a href="<?= $headline->url() ?>"><?= $headline->text() ?></a></li>
+      <?php endforeach ?>
+    </ol>
+
+  </nav>
+<?php endif ?>
+```
+
+### Result
+If you now add a `(toc)` placeholder to a page that contains headlines of the given level, you should see a ToC appear where you added the placeholder.
+
+## Alternative: A field method to generate the anchors
```

Comment:
I’d move this section up to the anchors. I first thought this was a ToC alternative, so I was confused reading this below the ToC section.